cijoe/scripts/latency_plotter.py

```diff
@@ -38,143 +38,179 @@ def data_as_a_function_of(data, x="iodepth", y="lat", filter=lambda _: True) ->
-                "xs": [],
-                "ys": [],
-                "ident": ident,
-                "label": label,
-                "group": groups[0],
-            }
-
-        samples[ident]["xs"].append(context[x])
-        samples[ident]["ys"].append(metrics[y])
-
-    return samples
-
 
-def data_as_fixed(data, x="iodepth", y="iops", xval=None):
-    """Organize and label data with 'y' points as a function of 'x' points"""
-
-    samples = {}
     for ident, metrics in data:
         context = metrics["ctx"]
-        if str(context[x]) != str(xval):
+
+        if not filter(metrics):
             continue
 
         if ident not in samples:
             label = context["name"]
-            groups = [
-                x.split("=")[1] for x in context["stem"].split("_") if "GROUP" in x
-            ]
             samples[ident] = {
-                "xs": [],
-                "ys": [],
+                "xys": [],
                 "ident": ident,
                 "label": label,
-                "group": groups[0],
+                "group": context["group"],
             }
 
-        samples[ident]["xs"].append(context[x])
-        samples[ident]["ys"].append(metrics[y])
+        samples[ident]["xys"].append((context[x], metrics[y]))
+
+    for ident in samples:
+        samples[ident]["xys"].sort()
 
     return samples
 
 
-def plot_attributes_from_step(step):
+def get_plot_attributes(step):
     """Load plot-attributes using paths in step"""
 
     limits_path = Path(step.get("with", {}).get("limits", "plot-limits.yaml"))
     legends_path = Path(step.get("with", {}).get("legends", "plot-legends.yaml"))
+    styles_path = Path(step.get("with", {}).get("styles", "plot-styles.yaml"))
 
     return {
         "limits": dict_from_yamlfile(limits_path),
         "legends": dict_from_yamlfile(legends_path),
+        "styles": dict_from_yamlfile(styles_path),
     }
 
 
-def build_plot_groups(data: Data):
-    """Group the `data`
+def draw_bar_plot(data, plot_attributes, xlabel=None, y_limit=None):
+    colors = plot_attributes["styles"].get("colors", ["#000000"])
+    hatches = plot_attributes["styles"].get("hatches", ["."])
+    width = 0.20
+
+    groups = list(set(map(lambda item: item["group"], data.values())))
+    is_grouped = len(groups) > 1
+
+    num_bars = len(data.items()) - 1
+    if is_grouped:
+        x_ticks = dict()
+        x_tick_info = dict([(group, [groups.index(group), 0]) for group in groups])
+    else:
+        unique_x_values = map(lambda xy: xy[0], list(data.values())[0]["xys"])
+        x_ticks = dict(
+            [(x, i + (num_bars * width) / 2) for i, x in enumerate(unique_x_values)]
+        )
+
+    if y_limit is None:
+        # assumes that the highest x-value will also have the highest y value.
+        # for each group, take y-value of the last (x,y) coordinate, and take the
+        # max value of these.
+        y_limit = max(map(lambda item: item["xys"][-1][1], data.values())) * 1.3
```

Comment:
Same comment as in the `bench_plotter`:

> This assumption can be avoided if you alter your lambda to give you the max y-value for each item.
Maybe you can do something like:
max(map(lambda item: max(item["xys"][:][1]), data.values())) * 1.3